make docs and examples use `istioctl kube-inject`

ayj opened the pull request:
I verified that demos/apps/bookinfo/README.md works with the `istioctl kube-inject` method. But I would be happy if anyone else wants to verify in their own setup before its merged.

I wasn't able to get doc/getting-started working, but it doesn't seem to work in my setup with unmodified changes after https://github.com/istio/istio/pull/86.

I'm not super happy about the need for `istio-inject` wrapper function but the alternative using `istioctl kube-inject` isn't great either, e.g.

```bash
 istioctl kube-inject \
             --initImage= hub=docker.io/istio/init:2017-03-22-17.30.06 \
             --runtimeImage= hub=docker.io/istio/runtime:2017-03-22-17.30.06
             -f <file> | kubectl apply -f -
```

We could this if/when we introduce a `stable` tag in which case we can update `kube inject` to use that by default and then usage would look like the following:

```bash
istioctl kube-inject -f <file> | kubectl apply -f -
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/istio/istio/87)
<!-- Reviewable:end -->

ayj commented:
cc @sebastienvas, it looks like some of the functions in https://github.com/istio/istio/blob/master/tests/kubeUtils.sh need to be updated to use `istioctl kube-inject` if we proceed with this approach. In particular, the `kubectl create` need to be updated to use `istioctl kube-inject` (see below). I'll wait to make this change until others review the concept.

```
istioctl kube-inject \
             --initImage= hub=docker.io/istio/init:2017-03-22-17.30.06 \
             --runtimeImage= hub=docker.io/istio/runtime:2017-03-22-17.30.06
             -f <file> | kubectl apply -f -
```

kyessenov commented:
What do you think about embedding SHA into istioctl and then using that for default docker image tags?

ayj commented:
That could work. Downside is users would need to download the latest version of istioctl anytime we identify a new set of stable images vs. making it implicit via versions specified in istio/istio/kubernetes/istioctl-kube-inject or a proper docker tag.

sebastienvas commented:
Does istioctl kube-inject support namespace for testing ? ie installing everything into one namespace and only listening on one namespace such that we can have multiple test running in one cluster that don t impact each other ?

ayj commented:
`istioctl kube-inject` doesn't do anything with namespaces. From the view of kubernetes there should be no difference between the hardcoded injection and kube-inject method.

frankbu commented on demos/apps/bookinfo/README.md:
```diff
@@ -35,35 +35,46 @@ rename the tool to `istioctl`. For example:
+
+   `istio-inject` is a wrapper function around `istioctl
+   kube-inject` which injects the istio runtime proxy into kubernetes
+   resources files. `istio-inject` and `istioctl kube-inject` are documented [here](https://github.com/istio/istio/blob/master/doc/istioctl.md#kube-inject).
```
better to use relative reference: ../../../doc/istioctl.md#kube-inject

frankbu commented on demos/apps/bookinfo/README.md:
````diff
@@ -35,35 +35,46 @@ rename the tool to `istioctl`. For example:
 
    ```bash
-   $ kubectl create -f bookinfo-istio.yaml
+   $ source ../../../kubernetes/istioctl-kube-inject.sh
````
Move this up before the previous step (cd) then it's only `source ./kubernetes/istioctl-kube-inject.sh`

frankbu commented on demos/apps/bookinfo/README.md:
````diff
@@ -35,35 +35,46 @@ rename the tool to `istioctl`. For example:
+   resources files. `istio-inject` and `istioctl kube-inject` are documented [here](https://github.com/istio/istio/blob/master/doc/istioctl.md#kube-inject).
+
+   ```bash
+   $ isito-inject bookinfo-istio.yaml | kubectl apply -f -
````
I'm not sure maintaining and using this istio-inject script is worth it. Maybe it would be better to just provide the whole `istioctl kube-inject ...`  command here instead. I think we can hardcode the hub (docker.io/istio) in the command and provide a simple grep/awk command to extract the correct tag from .../.../.../kubernetes/istio-install/istio-manager.yaml.

kyessenov replied:
I expect churn in the pod template we use, so we need to treat each SHA as incompatible with any other SHA. That's why I suggest we hardcode istioctl's SHA into docker image it uses by default. Once we have a stable alpha API, we can start putting alpha tag, and then switch to x.y.z versions.

ayj commented:
I created a new [PR](https://github.com/istio/manager/pull/388) to change `istioctl kube-inject` to use `--tag` and `--hub` flags instead of the full image name. I also changed the default images to point to `docker.io/istio/{init,runtime}:2017-03-22-17.30.06`.  With this change the separate the extra wrapper script shouldn't be necessary so I also removed that. Now the normal usage should look like this:

```bash
kubectl create -f <(istioctl kube-inject -f bookinfo-istio.yaml)
```

And using a new tag without updating to the latest istioctl version looks like this:

```bash
kubectl create -f <(istioctl kube-inject -f bookinfo-istio.yaml -tag <new-tag>)
```

frankbu commented:
This looks pretty good to me now. Couple of small issues:

1. now that the various yaml files are just plain app definitions, they should probably be renamed (hello-and-proxy.yaml -> hello.yaml, bookinfo-istio.yaml -> bookinfo.yaml, echo-app-proxy.yaml -> echo-app.yaml, etc.).
1. since the instructions say to use `kubectl create -f <(istioctl kube-inject -f bookinfo-istio.yaml)` to inject Istio, we need to emphasize that user MUST install and be using the "correct" version of `istioctl` (How does user confirm this?)

ayj commented:
@frankbu, re: point #2, user can confirm correct istioctl version with `version` subcomand, e.g.

```bash
$ istioctl version | grep GitRevision
GitRevision: <known-good-sha>
``` 

I'll update the docs with a specific revision once new istioctl binaries are uploaded. How is istioctl binary being released? It looks like @esnible uploaded the most recent set of binaries for win/osx/linux.

frankbu commented:
@ayj: Isn't there more to it than just making sure that the istioctl version is a known good one? Doesn't the istioctl version that we use need to match the tags being used by the version of istio/kuernetes/istio-install/istio-manager.yaml that we are using when starting the istio runtime services?

ayj commented:
s/\<known-good-sha\>/\<sha-that-corresponds-to-istio-manager-image\>

We can select a version of istioctol that corresponds to the image used into istio-manager.yaml. Ideally they would be the same once e2e acceptance test is automated, but for now we can select things manually, which I believe is what we've currently be doing anyway with the hardcoded  injection method.

sebastienvas commented:
@ayj I am planning to put istioctl binary in some istio-io bucket. And have them accessible from gcsweb.

esnible commented:
An extra line or two of --help for kube-inject with an example of piping to _kubectl apply_ would be very helpful.

Although the _istioctl kube-inject_ output looks correct after piping to _kubectl apply_ the `pod.alpha.kubernetes.io/init-containers` annotation appears twice.  Why?  To reproduce:

```
kubectl run tomcat --image=docker.io/library/tomcat
kubectl get deployment tomcat --output yaml | ./istioctl-osx kube-inject --filename - | kubectl apply -f -
kubectl get deployment tomcat --output yaml
```

ayj commented:
re:  annotation appearing twice, the same thing happens with the previous hardcoded yaml files. 

```bash
$ kubectl create -f hello-and-proxy.yaml
$ kubectl get deployment hello -o yaml | grep init-container
        pod.alpha.kubernetes.io/init-containers: '[{"name":"iptables","image":"docker.io/istio/runtime:2017-03-17-22.11.25","resources":{},"imagePullPolicy":"Always","securityContext":{"capabilities":{"add":["NET_ADMIN"]}}}]'
        pod.beta.kubernetes.io/init-containers: '[{"name":"iptables","image":"docker.io/istio/runtime:2017-03-17-22.11.25","resources":{},"imagePullPolicy":"Always","securityContext":{"capabilities":{"add":["NET_ADMIN"]}}}]'
```

From http://stackissue.com/kubernetes/kubernetes/move-init-container-feature-from-alpha-to-beta-31026.html 
> In 1.3, an init container is specified with this annotation key
> on the pod or pod template: `pods.alpha.kubernetes.io/init-containers`.
> 
> In 1.4, either that key or this key: pods.beta.kubernetes.io/init-containers`,
> can be used.
> 
> When you GET an object, you will see both annotation keys with the same values.

andraxylia commented on kubernetes/README.md:
```diff
@@ -34,9 +34,23 @@ start kubectl proxy, and edit Grafana's Istio-dashboard to use the proxy. Access
 ## Deploy your apps
 
+NOTE: Kubernetes admission controller for transparent proxy is not
+implementedyet . Use `istioctl kube-inject` to modify kubernetes
```
Insert  a space between implemented and yet.

istio-testing commented:
Jenkins job istio/presubmit passed

ayj commented:
Review status: 0 of 22 files reviewed at latest revision, 4 unresolved discussions.

---

*[demos/apps/bookinfo/README.md, line 59 at r1](https://reviewable.io:443/reviews/istio/istio/87#-KfxVFPOS4_kjt8C4DoY-r1-59:-Kg1XlfspF0tFMmI-xzX:b-896fix) ([raw file](https://github.com/istio/istio/blob/b6773eae00f8eb5a7ad3216f346acf2e4c4c3c8a/demos/apps/bookinfo/README.md#L59)):*
<details><summary><i>Previously, frankbu (Frank Budinsky) wrote…</i></summary><blockquote>

Move this up before the previous step (cd) then it's only `source ./kubernetes/istioctl-kube-inject.sh`
</blockquote></details>

Done.

---

*[demos/apps/bookinfo/README.md, line 64 at r1](https://reviewable.io:443/reviews/istio/istio/87#-KfxVFPOS4_kjt8C4DoY-r1-64:-Kg1XmOMO9IUkwsHuO08:b-896fix) ([raw file](https://github.com/istio/istio/blob/b6773eae00f8eb5a7ad3216f346acf2e4c4c3c8a/demos/apps/bookinfo/README.md#L64)):*
<details><summary><i>Previously, frankbu (Frank Budinsky) wrote…</i></summary><blockquote>

better to use relative reference: ../../../doc/istioctl.md#kube-inject
</blockquote></details>

Done.

---

*[demos/apps/bookinfo/README.md, line 67 at r1](https://reviewable.io:443/reviews/istio/istio/87#-KfxVFPOS4_kjt8C4DoY-r1-67:-Kg1XnIV_2C6yf7cqTln:b-896fix) ([raw file](https://github.com/istio/istio/blob/b6773eae00f8eb5a7ad3216f346acf2e4c4c3c8a/demos/apps/bookinfo/README.md#L67)):*
<details><summary><i>Previously, kyessenov (Kuat) wrote…</i></summary><blockquote>

I expect churn in the pod template we use, so we need to treat each SHA as incompatible with any other SHA. That's why I suggest we hardcode istioctl's SHA into docker image it uses by default. Once we have a stable alpha API, we can start putting alpha tag, and then switch to x.y.z versions.
</blockquote></details>

Done.

---

*[kubernetes/README.md, line 41 at r3](https://reviewable.io:443/reviews/istio/istio/87#-KfxVFPOS4_kjt8C4Dof-r3-41:-Kg1Xo-9OzFO_i_L_qdi:b-896fix) ([raw file](https://github.com/istio/istio/blob/bb493d0cd7c9167501a56f05f184693289771b72/kubernetes/README.md#L41)):*
<details><summary><i>Previously, andraxylia (Andra Cismaru) wrote…</i></summary><blockquote>

Insert  a space between implemented and yet.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/istio/istio/87)*
<!-- Sent from Reviewable.io -->

istio-testing commented:
Jenkins job istio/presubmit passed

ayj commented:
Pre-submit is now passing with `kube-inject` injecting proxy into bookinfo.yaml. I pushed new versions of `istioctl` binary which should work for now until those can be pulled from GCS bucket. 

@frankbu 's comment about mismatched istioctl is noted, so perhaps we wait and see if its a problem and react accordingly? Other than that I think all comments have been addressed. If not, please let me know.

frankbu commented:
@ayj: fine with me to merge. I'd like to update the README.md to show both before and after (istio-inject) diagrams of the bookinfo components, but if you merge this PR, I can do that as a separate PR afterwords.
